chainerx_cc/chainerx/cuda/cuda_device/indexing.cu

```diff
@@ -185,58 +195,64 @@ void TakeImpl(Device& device, const Array& a, const Array& indices, int8_t axis,
-        Axes out_perm = MakeRollingPermutation(axis, axis + indices.ndim(), out.ndim());
-        out_iarray.Permute(out_perm);
-        Shape out_shape = internal::TransposeShape(out.shape(), out_perm);
-        Indexer<> out_indexer{out_shape};
-
-        IndexableArray<const TIndex> indices_iarray{indices};
-        Indexer<> indices_indexer{indices.shape()};
-
-        // size of (Ni..., Nj...) part
-        TIndex common_total_size = gsl::narrow<TIndex>(a_indexer.total_size() / a_shape[0]);
-
-        TIndex axis_dim = gsl::narrow<TIndex>(a_shape[0]);
-
-        // TODO(niboshi): Calculate kMaxBlockSize per device
-        std::lock_guard<std::mutex> lock{*cuda_internal::g_mutex};
-        int64_t total_size = out_indexer.total_size();
-        static const int kWrapMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeWrapCudaKernel<T, TIndex>).block_size;
-        static const int kClipMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeClipCudaKernel<T, TIndex>).block_size;
-        int64_t grid_size;
-        int64_t block_size;
-        switch (mode) {
-            case IndexBoundsMode::kRaise:
-                throw BackendError{"Take with mode='raise' is not supported with CUDA backend"};
-            case IndexBoundsMode::kDefault:
-            case IndexBoundsMode::kWrap:
-                grid_size = (total_size + kWrapMaxBlockSize - 1) / kWrapMaxBlockSize;
-                block_size = std::min<TIndex>(total_size, kWrapMaxBlockSize);
-                TakeWrapCudaKernel<<<grid_size, block_size>>>(
-                        a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
-                break;
-            case IndexBoundsMode::kClip:
-                grid_size = (total_size + kClipMaxBlockSize - 1) / kClipMaxBlockSize;
-                block_size = std::min<TIndex>(total_size, kClipMaxBlockSize);
-                TakeClipCudaKernel<<<grid_size, block_size>>>(
-                        a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
-                break;
+        TIndex num_indices = indices.GetTotalSize();
+        TIndex target_dim = a.shape()[axis];
+        TIndex num_iters = out.GetTotalSize();
+        TIndex right_dim = std::accumulate(a.shape().begin() + axis + 1, a.shape().end(), int64_t{1}, std::multiplies<>());
+        static const int k1DMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeCudaKernel<T, TIndex, 1>).block_size;
+        static const int kNDMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeCudaKernel<T, TIndex, kDynamicNdim>).block_size;
+        Shape new_shape_a = collapse_nd_array(a);
+        Shape new_shape_i = collapse_nd_array(indices);
+        Shape new_shape_out = collapse_nd_array(out);
+        Array af = a.Reshape(new_shape_a);
+        Array indicesf = indices.Reshape(new_shape_i);
+        Array outf = out.Reshape(new_shape_out);
```

Comment:
It is guaranteed that no memory copy is performed in these `Reshape`. How about directly using `internal::MakeArray` to avoid redundant checks for memory copy?